compiler/sem/mirexec.nim

```diff
@@ -0,0 +1,618 @@
+    of opFork, opGoto, opLoop:
+      dest: JoinId
+    of opJoin: # used for two things:
+               # 1. associating a goto/fork target with a node
+               # 2. to know where control-flow rejoins during reverse execution
+      id: JoinId
+
+  ControlFlowGraph* = object
+    instructions: seq[Instr]
+    map: seq[InstrPos] ## joind ID -> instruction
+
+  TraverseState* = object
+    exit*: bool     ## used to communicate to ``traverse`` that the active path
+                    ## should be "killed" (not followed further). Reset to `false`
+                    ## whenever control is passed back to the iterator.
+                    ## When traversal is finished, `exit` is set to indicates
+                    ## whether traversal reached the end of the given span
+                    ## (`true`) or not (`false`)
+    escapes*: bool  ## whether control-flow escapes the given span
+
+  LoopEntry = tuple
+    start: JoinId ## the join point that marks the start of the loop
+    fin: InstrPos ## the 'loop' instruction
+    isRun: bool   ## whether the loop is currently processed
+
+  ExecMask = uint8
+
+  ExecState = object
+    ## Execution environment state for abstract reverse execution
+    visited: seq[ExecMask]
+      ## the execution mask associated with each join point
+      # XXX: since an execution only occupies 2 bits, it would make sense to
+      #      use a a packed sequence here
+    loops: seq[LoopEntry]
+      ## the loop stack. Remembers the loops the current basic block is
+      ## located inside
+
+    state: ExecMask
+      ## the active execution mask
+
+    pc: InstrPos
+      ## the program counter. Points to the CFG instruction that is executed
+      ## next
+    i: NodePosition
+      ## points to the next item to yield
+    inLoop: int
+      ## the number of currently processed loops. Note that this is not
+      ## necessarily the same as ``loops.len``
+
+# TODO: copied from ``ast_query.nim``. Either export the original or use a
+#       different approach - duplicating the constant is not acceptable
+const magicsThatCanRaise = {
+  mNone, mSlurp, mStaticExec, mParseExprToAst, mParseStmtToAst, mEcho}
+
+func incl[T](s: var seq[T], v: sink T) =
+  ## If not present already, adds `v` to the sorted ``seq`` `s`
+  var i = 0
+  while i < s.len and s[i] < v:
+    inc i
+
+  if i >= s.len or s[i] != v:
+    s.insert(v, i)
+
+func compare(a: Instr, b: NodePosition): int =
+  ord(a.node) - ord(b)
+
+template lowerBound(c: ControlFlowGraph, node: NodePosition): InstrPos =
+  lowerBound(c.instructions, node, compare).InstrPos
+
+template upperBound(c: ControlFlowGraph, node: NodePosition): InstrPos =
+  upperBound(c.instructions, node, compare).InstrPos
+
+func `[]`(c: ControlFlowGraph, pc: SomeInteger): lent Instr {.inline.} =
+  c.instructions[pc]
+
+func computeCfg*(tree: MirTree): ControlFlowGraph =
+  ## Computes the control-flow graph for the given `tree`. This is a very
+  ## expensive operation!
```

Comment:
```suggestion
  ## Computes the control-flow graph for the given `tree`. This is a very
  ## expensive operation! The high cost is due to two essential reasons:
  ##
  ## 1. a control flow graph needs to materialize all paths for a given `tree`
  ## 2. the amount of allocations/bookkeeping necessary to do so
```

This is partly me just confirming my understanding, the first point I'm fairly confident about and the second one is more me guessing by looking at the push/pop of sequences and the `Table` in `ClosureEnv`.

Aside, I spent a few minutes trying to reason about replacing that table and the best I could figure was maybe the following:

```
type
  ClosureEnv = object
    # ... existing bits except joins
    joinedNodes: seq[NodePosition]
    nodeIsJoined: PackedSet[NodePosition]`

proc addJoin(env: var ClosureEnv, dest: NodePosition): JoinPoint =
  let (id, isNew) =
    if env.nodeIsJoined.containsOrIncl(dest):
      env.nodeIsJoined.add dest
      (env.joinedNodes.len - 1, true)
    else:
      (env.joinedNodes.find dest, false)  # likely more efficient with a reverse find

  JoinPoint(id: id, node: dest, isNew: isNew)
```

This is mostly a test of my understanding and a bit of fun, I doubt it's an improvement.

Reply:
I wouldn't say that the control-flow graph needs to materialize all paths for `tree`, rather only all edges (in the form of instructions). However, I might be misunderstanding what you mean.

Regarding the reason for why `computeCfg` is expensive, the last time I profiled it, the by far largest amount of time was spent inside the tree traversal routines. For a general perspective, using a `-d:release` compiler built with `--gc:orc` to compile itself with `--gc:orc` as the benchmark, more time (~15%) is spent in `computeCfg` than in `mirgen` (i.e. `genProcBody`).

I did another profiling session using the newest commit -- the six most time consuming calls inside `computeCfg` are:
| Procedure | Time spent |
| --------- | ---------- |
| `mirtrees.findEnd` | 0.401 % |
| `mirtrees.parent` | 0.224 % |
| `algorithm.sort` | 0.0476 % |
| `commit` | 0.0174 % |
| `mirtrees.findParent` | 0.0174 % |
| `addEdge` | 0.011 % |

The "Time spent" is a percentage of the total compilation time. Using either recursion or a skip-list (or both) could reduce the amount of time spent on tree traversal.

---

The snippet you posted is, apart from two problems that I believe are unintended, correct. I was curious and ran a comparison between your code and the current implementation. Instrumenting a release compiler that is booted with `--gc:orc` and using it to compile the `devel` compiler with `--gc:orc` yields:

|   | Time taken (in milliseconds) | Allocations | Deallocations |
| - | ---------------------------- | ----------- | ------------- |
| `Table` | 151 | 35219 | 17779 |
| `PackedSet` + `seq` + `find` | 152 - 153 | 36865 | 19425 |
| `PackedSet` + `seq` + `rfind` | 154 - 155 | 36865 | 19425 |

Reply:
The allocations really dictate performance, kinda sad that `PackedSet`s weren't able to get ahead. I'll have to look closer especially once the AST starts to get the DOD treatment as I'd been hoping they'd do a lot of heavy lifting in a few areas.